substrate/frame/staking/src/slashing.rs

```diff
@@ -341,31 +333,24 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
 
 		let validator_index_u32 = validator_index as u32;
 
-		match offending.binary_search_by_key(&validator_index_u32, |(index, _)| *index) {
-			// this is a new offending validator
-			Err(index) => {
-				offending.insert(index, (validator_index_u32, disable));
-
-				let offending_threshold =
-					T::OffendingValidatorsThreshold::get() * validators.len() as u32;
-
-				if offending.len() >= offending_threshold as usize {
-					// force a new era, to select a new validator set
-					<Pallet<T>>::ensure_new_era()
-				}
-
-				if disable {
-					T::SessionInterface::disable_validator(validator_index_u32);
-				}
-			},
-			Ok(index) => {
-				if disable && !offending[index].1 {
-					// the validator had previously offended without being disabled,
-					// let's make sure we disable it now
-					offending[index].1 = true;
-					T::SessionInterface::disable_validator(validator_index_u32);
-				}
+		let DisablingDecision { disable_offenders } = T::DisablingStrategy::make_disabling_decision(
```

Comment:
As per a comment before, I'd remove the `DisablingDecision` struct and just get the `disable_offenders` vec directly here.

Reply:
(same for `DisablingDecisionContext` below)

Reply:
Done